go-controller/pkg/ovn/dns-name-resolver/egressfirewall_external_dns.go

```diff
@@ -0,0 +1,386 @@
+	extEgDNS := &ExternalEgressDNS{
+		dnsNames:          make(map[string]dnsResolvedName),
+		namespaceDNSNames: make(map[string]sets.Set[string]),
+		addressSetFactory: addressSetFactory,
+		controllerName:    controllerName,
+		dnsNameToResolver: make(map[string]string),
+		resolverToDNSName: make(map[string]string),
+	}
+	extEgDNS.initControllers(watchFactory)
+
+	return extEgDNS
+}
+
+// initControllers initializes the controllers for the different resources
+func (extEgDNS *ExternalEgressDNS) initControllers(watchFactory *factory.WatchFactory) {
+	dnsSharedIndexInformer := watchFactory.DNSNameResolverInformer().Informer()
+	extEgDNS.dnsLister = ocpnetworklisterv1alpha1.NewDNSNameResolverLister(dnsSharedIndexInformer.GetIndexer())
+	dnsConfig := &controller.Config[ocpnetworkapiv1alpha1.DNSNameResolver]{
+		RateLimiter:    workqueue.NewItemFastSlowRateLimiter(time.Second, 5*time.Second, 5),
+		Informer:       dnsSharedIndexInformer,
+		Lister:         extEgDNS.dnsLister.List,
+		ObjNeedsUpdate: dnsNeedsUpdate,
+		Reconcile:      extEgDNS.reconcileDNSNameResolver,
+	}
+	extEgDNS.dnsNameResolverController = controller.NewController[ocpnetworkapiv1alpha1.DNSNameResolver]("dnc-dns-controller", dnsConfig)
+}
+
+// dnsNeedsUpdate returns true if a dns name resolver object is either added
+// or deleted. If a dns name resolver is updated, then dnsNeedsUpdate returns
+// true if the .status of the object is modified.
+func dnsNeedsUpdate(oldObj, newObj *ocpnetworkapiv1alpha1.DNSNameResolver) bool {
+	if oldObj == nil || newObj == nil {
+		return true
+	}
+	return !reflect.DeepEqual(oldObj.Status, newObj.Status)
+}
+
+// reconcileDNSNameResolver reconciles a DNSNameResolver object. If an object was deleted,
+// then egressFirewallExternalDNS.Delete is called for the corresponding DNS name. If an
+// object is added/updated, then egressFirewallExternalDNS.Add is called for the
+// corresponding DNS name along with the associated addresses.
+func (extEgDNS *ExternalEgressDNS) reconcileDNSNameResolver(key string) error {
+	// Split the key in namespace and name of the corresponding object.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("reconcileDNSNameResolver failed to split meta namespace cache key %s for dns name resolver: %v", key, err)
+		return nil
+	}
+	// Fetch the dns name resolver object using the name and namespace.
+	obj, err := extEgDNS.dnsLister.DNSNameResolvers(namespace).Get(name)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// DNSNameResolver object was deleted.
+
+			// Get the corresponding DNS name, if avialable.
+			dnsName, found := extEgDNS.resolverToDNSName[name]
+			// If the DNS name was not found then return.
+			if !found {
+				return nil
+			}
+
+			// Delete the address set associated with the DNS name.
+			var err error
+			waitErr := wait.PollUntilContextTimeout(context.Background(), 100*time.Millisecond, 1*time.Second, true, func(ctx context.Context) (bool, error) {
```

Comment:
why do you prefer an extra retry logic instead of just returning an error and letting the controller retry?

Reply:
The unit test fails without the wait. The `gomega.Eventually()` doesn't wait for the next retry of the controller and fails the test.

Reply:
that sounds like we should change the test and the implementation? let me see what that test is doing, maybe i will have some ideas